### ISSUE TITLE
Introduce a way to manage focus on the RichText editor

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#9740]
See discussion on Slack: https://wordpress.slack.com/archives/C02QB2JS7/p1536570671000100

In 4.0 the RichText `onSetup` prop will be deprecated and changed to `unstableOnSetup`. To my understanding, `unstableOnSetup` won't be removed so soon, see https://github.com/WordPress/gutenberg/pull/9106, because it's still necessary in current implementations of core blocks. 

However, in the future there will still be the need to access the editor instance. Simple use case: managing focus. Imagine a custom block that uses one or more RichText components and needs to manage focus programmatically depending on specific workflows.

As mentioned in the Slack discussion, one option could be introducing a `focus` function to the `RichText` component. Definitely not my area of expertise, there are probably other options and considerations to explore. /Cc @iseulde @aduth

**Comment by iseulde** [comment]
Maybe `onSelect`?

Additionally there's no way currently to set the focus at the end of the field.

**Comment by iseulde** [comment]
I think this needs to be addressed before API freeze. I tried to make a block with multiple RichText fields and it is impossible to manage focus, even with `unstableOnFocus`. Passing `isSelected={ true }` does nothing to focus the field if requested.

**Comment by iseulde** [comment]
Plus we are still using `unstableOnFocus` in some blocks.

**Comment by aduth** [comment]
>Imagine a custom block that uses one or more RichText components and needs to manage focus programmatically depending on specific workflows.

>I tried to make a block with multiple RichText fields and it is impossible to manage focus

The use-cases here are quite vague, or I'm not sufficiently imaginative. What is such a block trying to achieve with programmatically managing focus?

**Comment by iseulde** [comment]
E.g. When you have multiple RichText fields in a block and you'd like to manage split and merge behaviour. Currently this is all possible, except for setting focus to the correct field. The block in question I was working on is the list blocks (with multiple RichText fields), but I can imaging this being useful for many kinds of blocks with Rich input fields.

**Comment by iseulde** [comment]
At the very least I would like to see our own blocks not using `unstableOnFocus`.

**Comment by iseulde** [comment]
Or maybe it should be kept (call on `contentEditable` focus for a stable result) and be renamed to e.g. `onSelect` which could go together with `isSelected` to set focus on the field.

This could also be revised as part of the focus setting we do currently in `compontentDidUpdate`, which is not really the greatest logic we have in `RichText`. :/

**Comment by iseulde** [comment]
> Related: #6419

I tried to use `setFocusedElement` where we currently use `unstableOnFocus` in the table block, but that didn't seem to work. Cc @gziolo.

**Comment by gziolo** [comment]
> I tried to use setFocusedElement where we currently use unstableOnFocus in the table block, but that didn't seem to work.

Yes, it was a quite complex use case. I need to refresh my memory to figure out what was blocking us from using `setFocusedElement`.

**Comment by gziolo** [comment]
This is my previous attempt #6871 to fix it without using `onFocus` prop which was renamed later to `unstableOnFocus` as we decided to avoid landing this PR. It still needs some tinkering how to make it work in a more automated way. I think the biggest blocker was mentioned by @youknowriad in https://github.com/WordPress/gutenberg/pull/6871#issuecomment-392895193:
> The issue is that sometimes RichText loses the focus but we still want to keep it selected. (Link modal, toolbar buttons), It's probably doable but not simple because of the fact that popovers show outside the RichText's node.

**Comment by afercia** [comment]
Quoting from the duplicate issue #12587 

> Looks like with the removal of `unstableOnSetup` in https://github.com/WordPress/gutenberg/pull/10744 there isn't a built-in way to manage focus on the RichText editable area any longer.

> Managing focus is essential in some flows, especially for custom block types with multiple RichText instances. While the removal of `unstableOnSetup` was expected to happen without notice, a new, alternative, method was expected as well. 

> Other editors expose a `focus` method: for example, Draft.js explains very well why this is needed, see https://draftjs.org/docs/advanced-topics-managing-focus

> Discussed a bit this issue with our technical team at Yoast and we'd tend to think that also all use-cases that were offered by a reference to TinyMCE would need to be offered by the wrapper, `RichText` in this case.

**Comment by atimmer** [comment]
I think this is essential functionality for an editable field component, which RichText is. I understand the choice to not expose the TinyMCE instance, but focusing an editable field is core functionality.

**Comment by hypest** [comment]
👋 , with https://github.com/WordPress/gutenberg/pull/14640 the selection logic has received a significant update, including how components (RichText in particular) get focused. I wonder if this ticket needs some updating so we better understand the current state of things. Wdyt @ellatrix ?

**Comment by chrisvanpatten** [comment]
Does this relate to the situation where a RichText component within a block receives focus when a block is selected? I'd love to have a flag to prevent that behavior (so the user has to explicitly select the RichText area to focus it) but I'm not sure if that counts as part of this issue, or in another issue :)

**Comment by afercia** [comment]
No this is unrelated. It's about introducing a way to set focus _programmatically_ on the contenteditable. Currently, there's no way to get a block editor (TinyMCE) instance, see original description and linked references.

**Comment by talldan** [comment]
A related issue for the table block is potentially https://github.com/WordPress/gutenberg/issues/14904. I can think of a workaround using event handlers and `element.focus()`, but it'd be great if there was a way to forward focus from the surrounding td to the child `RichText`.

At some point I also want to introduce the ability to navigate the table block using arrow keys, which might be another use case, but also something that could use `element.focus()` as an intermediate step.

**Comment by afercia** [comment]
Update: current state of the RichText component with regards to focus management after one year this issue was created:
- `unstableOnSetup` was removed 
- `setFocusedElement` is now deprecated, see https://github.com/WordPress/gutenberg/pull/17421 and the recommended alternative is `selection state from the block editor store`
- `isSelected` is now `__unstableIsSelected`, not sure how this is impacting / will impact the ability to manage focus

To recap the nature of this issue: seems to me all these changes fit only with the internal needs of Gutenberg. They don't take into consideration custom usages of RichText outside of the Gutenberg ones. Instead, plugins may need to implement their own focus management mechanisms.

Consider for example a block that uses multiple RichText components like the one below: each highlighted area is a RichText:

<img width="776" alt="Screenshot 2019-09-13 at 12 08 02" src="https://user-images.githubusercontent.com/1682452/64855507-51bab980-d620-11e9-9e72-036102b4c8e1.png">

One year ago, managing focus within such a block was pretty simple. Unfortunately, all the changes in the last year made it more and more challenging. At the point that providing basic keyboard accessibility is now very hard.

To me, it appears clear that Gutenberg should provide a mechanism to manage focus on the RichText instances within a block. Gutenberg removed features that made this kind of focus management possible: now, it's a Gutenberg responsibility to provide equivalent features.

**Comment by ellatrix** [comment]
@afercia Looking into focus and rich text right now as I'm refactoring things internally. What do you need?

1. There's currently a way to know when rich text receives focus through `unstableOnFocus`, which will be made stable with the event exposed.
2. You can set focus by passing a ref to rich text and calling `ref.current.focus()` (the normal [element method](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/focus)).

What more do you need?

If there's anything more, could you update the issue description or create a new issue? It's a bit outdated now.

**Comment by afercia** [comment]
@mtias for history and documentation, can you please add a comment pointing to where this issue was fixed?
I think it's important in an open source project to clarify why an issue gets closed, much like it's expected [on Trac](https://make.wordpress.org/core/handbook/contribute/trac/), where the 'resolution' status and a comment are pretty the norm when closing a ticket.

**Comment by emeaguiar** [comment]
I am also curious to know how this is resolved now, do we have a way to set focus to a RichText component?

**Comment by mtias** [comment]
Given @ellatrix articulated how to focus rich text and there were no further comments on what else might be needed closing seemed appropriate to help keep the repository in shape.

**Comment by afercia** [comment]
> Given @ellatrix articulated how to focus rich text and there were no further comments on what else might be needed closing seemed appropriate to help keep the repository in shape.

I'm not sure the feedback from @ellatrix is up to date, given `unstableOnFocus` is now only used in a few `native` files. At least, that's my understanding. I don't see any documentation about `unstableOnFocus` as well.

Anyways, the point is more about process. This is an open source project. Please don't close issues without a comment to explain why you are closing it. Thanks.

**Comment by ellatrix** [comment]
I believe onFocus has been made stable a long time ago and works like any other React event
